### ISSUE TITLE
Compile error on Alpine Linux, Minetest 0.4.13

### ISSUE THREAD

**Issue by iio7** [minetest/minetest#3124]
[ 75%] Building CXX object src/CMakeFiles/minetestserver.dir/gettext.cpp.o
[ 75%] Building CXX object src/CMakeFiles/minetestserver.dir/httpfetch.cpp.o
In file included from /usr/include/curl/curl.h:2343:0,
                 from /home/minetest/minetest-0.4.13/src/httpfetch.cpp:121:
/usr/include/curl/multi.h:146:40: error: 'fd_set' has not been declared
                                        fd_set *read_fd_set,
                                        ^
/usr/include/curl/multi.h:147:40: error: 'fd_set' has not been declared
                                        fd_set *write_fd_set,
                                        ^
/usr/include/curl/multi.h:148:40: error: 'fd_set' has not been declared
                                        fd_set *exc_fd_set,
                                        ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp: In member function 'void CurlFetchThread::waitForIO(long int)':
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:558:3: error: 'fd_set' was not declared in this scope
   fd_set read_fd_set;
   ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:558:3: note: suggested alternative:
In file included from /usr/include/fortify/sys/select.h:4:0,
                 from /usr/include/sys/time.h:9,
                 from /home/minetest/minetest-0.4.13/src/porting.h:229,
                 from /home/minetest/minetest-0.4.13/src/httpfetch.cpp:21:
/usr/include/sys/select.h:25:3: note:   'porting::fd_set'
 } fd_set;
   ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:559:10: error: expected ';' before 'write_fd_set'
   fd_set exc_fd_set;                                                                                                [0/1074]
          ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:563:18: error: aggregate 'CurlFetchThread::waitForIO(long int)::timeval select_tv' has incomplete type and cannot be defined
   struct timeval select_tv;
                  ^
In file included from /usr/include/fortify/sys/select.h:4:0,
                 from /usr/include/sys/time.h:9,
                 from /home/minetest/minetest-0.4.13/src/porting.h:229,
                 from /home/minetest/minetest-0.4.13/src/httpfetch.cpp:21:
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:566:12: error: 'read_fd_set' was not declared in this scope
   FD_ZERO(&read_fd_set);
            ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:567:12: error: 'write_fd_set' was not declared in this scope
   FD_ZERO(&write_fd_set);
            ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:568:12: error: 'exc_fd_set' was not declared in this scope
   FD_ZERO(&exc_fd_set);
            ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:570:37: error: 'read_fd_set' was not declared in this scope
   mres = curl_multi_fdset(m_multi, &read_fd_set,
                                     ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:571:6: error: 'write_fd_set' was not declared in this scope
     &write_fd_set, &exc_fd_set, &max_fd);
      ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:571:21: error: 'exc_fd_set' was not declared in this scope
     &write_fd_set, &exc_fd_set, &max_fd);
                     ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:599:17: error: 'select' was not declared in this scope
       &select_tv);
                 ^
/home/minetest/minetest-0.4.13/src/httpfetch.cpp:599:17: note: suggested alternative:
In file included from /usr/include/fortify/sys/select.h:4:0,
                 from /usr/include/sys/time.h:9,
                 from /home/minetest/minetest-0.4.13/src/porting.h:229,
                 from /home/minetest/minetest-0.4.13/src/httpfetch.cpp:21:
/usr/include/sys/select.h:32:5: note:   'porting::select'
 int select (int, fd_set *__restrict, fd_set *__restrict, fd_set *__restrict, struct timeval *__restrict);
     ^
src/CMakeFiles/minetestserver.dir/build.make:2285: recipe for target 'src/CMakeFiles/minetestserver.dir/httpfetch.cpp.o' failed
make[2]: *** [src/CMakeFiles/minetestserver.dir/httpfetch.cpp.o] Error 1
make[2]: *** Waiting for unfinished jobs....
CMakeFiles/Makefile2:158: recipe for target 'src/CMakeFiles/minetestserver.dir/all' failed
make[1]: *** [src/CMakeFiles/minetestserver.dir/all] Error 2
Makefile:136: recipe for target 'all' failed
make: *** [all] Error 2

**Comment by PilzAdam** [comment]
Your system seems broken; there are errors in the curl header files, we can't do anything about it.

**Comment by est31** [comment]
While its much better to run minetest with curl support (for multiplayer), and fixing the curl headers, you can try to disable curl with passing `-DENABLE_CURL=0` to cmake.

**Comment by sfan5** [comment]
This is probably caused by Alpine Linux using musl libc

**Comment by ncopa** [comment]
Which version of alpine and curl was it? I am trying to fix curl but am not able to create a minimal testcase where it can be reproduced.

**Comment by iio7** [comment]
It is Alpine 3.2.1 and Curl 7.42.1.

**Comment by ncopa** [comment]
I was able to reproduce. This does look like curl headers are broken, but I was not able to reproduce without minetest. I also found out that if I include curl.h before porting.h (from minetest) then it no longer fails.

I suspect that something in porting.h or something that is pulled by porting.h defines or undefines something that confuses curl.h.

I am not convinced that the error is in curl headers. If it is, then I don't know how to trigger it.
